cedar-lean/Cedar/Thm/Data/List.lean

```diff
@@ -841,73 +941,323 @@ theorem forall₂_implies_all_right {α β} {R : α → β → Prop} {xs : List
+    replace ⟨⟨xhd, h₁⟩, h₂⟩ := h₁
+    replace ⟨xtl, ih⟩ := ih h₂
+    exists xhd :: xtl
+    simp [h₁, ih, pure, Except.pure]
+
+theorem all_from_ok_implies_mapM_ok {α β γ} {f : α → Except γ β} {ys : List β} :
+  (∀ y ∈ ys, ∃ x, f x = .ok y) →
+  ∃ xs, List.mapM f xs = .ok ys
+:= by
+  intro h
+  have ⟨xs, h₂⟩ := all_from_ok_implies_mapM'_ok h
+  rw [List.mapM'_eq_mapM] at h₂
+  exists xs
 
+theorem mapM'_some_iff_forall₂ {α β} {f : α → Option β} {xs : List α} {ys : List β} :
+  List.mapM' f xs = .some ys ↔
+  List.Forall₂ (λ x y => f x = .some y) xs ys
+:= by
+  constructor
+  case mp =>
+    intro h₁
+    induction xs generalizing ys
+    case nil =>
+      simp only [mapM'_nil, pure, Option.some.injEq] at h₁
+      subst h₁
+      exact List.Forall₂.nil
+    case cons xhd xtl ih =>
+      simp only [mapM'_cons, pure, Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq] at h₁
+      replace ⟨yhd, h₁, ytl, h₂, h₃⟩ := h₁
+      subst h₃
+      exact List.Forall₂.cons h₁ (ih h₂)
+  case mpr =>
+    intro h₁
+    induction xs generalizing ys
+    case nil =>
+      simp only [forall₂_nil_left_iff] at h₁
+      simp only [mapM'_nil, pure, Except.pure, h₁]
+    case cons xhd xtl ih =>
+      simp only [mapM'_cons, pure, Except.pure]
+      replace ⟨yhd, ytl, h₁, h₃, h₄⟩ := forall₂_cons_left_iff.mp h₁
+      subst ys
+      cases h₂ : f xhd
+      case none => simp [h₁] at h₂
+      case some y' =>
+        simp only [h₁, Option.some.injEq] at h₂
+        subst y'
+        simp [@ih ytl h₃]
+
+theorem mapM_some_iff_forall₂ {α β} {f : α → Option β} {xs : List α} {ys : List β} :
+  List.mapM f xs = .some ys ↔
+  List.Forall₂ (λ x y => f x = .some y) xs ys
+:= by
+  rw [← List.mapM'_eq_mapM]
+  exact mapM'_some_iff_forall₂
+
+/--
+  Note that the converse is not true:
+  counterexample `xs` is `[1]`, `ys` is `[1, 2]`, `f` is `Option.some`
+-/
 theorem mapM'_some_implies_all_some {α β} {f : α → Option β} {xs : List α} {ys : List β} :
   List.mapM' f xs = .some ys →
   ∀ x ∈ xs, ∃ y ∈ ys, f x = .some y
 := by
   intro h
-  exact forall₂_implies_all_left (mapM'_some_implies_forall₂ h)
+  exact forall₂_implies_all_left (mapM'_some_iff_forall₂.mp h)
+
+theorem mapM_some_implies_all_some {α β} {f : α → Option β} {xs : List α} {ys : List β} :
+  List.mapM f xs = .some ys →
+  ∀ x ∈ xs, ∃ y ∈ ys, f x = .some y
+:= by
+  rw [← List.mapM'_eq_mapM]
+  exact mapM'_some_implies_all_some
 
+/--
+  Note that the converse is not true:
+  counterexample `ys` is `[1]`, `xs` is `[1, 2]`, `f` is `Option.some`
+-/
 theorem mapM'_some_implies_all_from_some {α β} {f : α → Option β} {xs : List α} {ys : List β} :
   List.mapM' f xs = .some ys →
   ∀ y ∈ ys, ∃ x ∈ xs, f x = .some y
 := by
   intro h
-  exact forall₂_implies_all_right (mapM'_some_implies_forall₂ h)
+  exact forall₂_implies_all_right (mapM'_some_iff_forall₂.mp h)
+
+theorem mapM_some_implies_all_from_some {α β} {f : α → Option β} {xs : List α} {ys : List β} :
```

Comment:
Why the "from" in the lemma name?

Reply:
> While the `*_implies_all_some` lemmas state that every element of the input list has a corresponding element in the output list, the `*_implies_all_from_some` lemmas state that every element of the output list has a corresponding element in the input list.

Yes, that was the rationale:  input / output distinction.